alacritty/src/event.rs

```diff
@@ -472,6 +498,15 @@ impl<N: Notify + OnResize> Processor<N> {
                 );
             }
 
+            #[cfg(not(any(target_os = "macos", windows)))]
+            {
+                if event_loop.is_wayland()
+                    && !self.display.window.should_draw.load(Ordering::Relaxed)
+                {
+                    return;
+                }
+            }
```

Comment:
What you seem to do in this PR is basically just ignore redraw opportunities until Wayland says it's ready, but I'm a bit concerned if that's the ideal approach.

At least in theory, that should open us for potential latency issues. If we miss the vsync just by a hair and go into the next event loop, every event we receive leaves us open to miss the vsync window. Which would mean that we'd probably miss the vsync and then latency is increased by one frame whenever this happens.

It's possible that this still improves things in general, but this seems like a bit of a design flaw? Have you considered this and potential solutions for it?

Reply:
>At least in theory, that should open us for potential latency issues. If we miss the vsync just by a hair and go into the next event loop, every event we receive leaves us open to miss the vsync window. Which would mean that we'd probably miss the vsync and then latency is increased by one frame whenever this happens.

You're relying on when to draw already, since vsync doesn't exists on Wayland at all. Compositors are scheduling frames, in that case they are sending you event to redraw in some before the next vblank, e.g.  max_render_time on sway, weston is using 7ms (Could be changed). If you can't render under the time they are giving to you that's really unfortunate.

Ideally you just put your rendering code into a callback, however that's not something possible sometimes, and some apps are doing the same approach, like Firefox.

The only way you can miss redraw in that case, if you really can't render under the time your compositor is giving to you, but it's not our problem, if it can't schedule frames properly. Compositor can collect statics on how you reply to its redraw requests and tune its scheduling of frame callbacks for each client, so if you start to draw under 13ms or you'll be missing times, it can start sending frame events earlier. So I don't see any issue in doing it this way, and if client is missing redraw, it's compositor's fault.

I've been running this PR for a long time already on slow and a bit faster system, and latency on it is way lower. I've tested weston and sway.

>Have you considered this and potential solutions for it?

Up to compositor.

Reply:
To summarize our discussion on IRC:

Sway's latency is delayed by an entire vblank interval by default, but can be lowered to any value in the config according to the documentation:

https://github.com/swaywm/sway/blob/58703e4f24ae2738f9459c0c092f41a827be007e/sway/sway.5.scd#L188-L202 

So this means that we can't really miss the vblank, since we have up to an entire frame ahead of it and do not need to render immediately after the callback came in. This is also necessary for us since there might be pending resize events which need to be processed.

Of course if a compositor chose to, they could break this. But that's a compositor's problem not the application's. Sway plays it safe here by basically acting like our X backend by default, but giving the user the choice to optimize it themselves.

So this should give us a great balance of leaving as much time as possible before rendering to reduce input latency and making it in time for the next vblank.